src_c/draw.c

```diff
@@ -983,27 +975,75 @@ set_at(SDL_Surface *surf, int x, int y, Uint32 color)
     return 1;
 }
 
-#define DRAWPIX32(pixel, colorptr, br, blend)                               \
-    if (blend) {                                                            \
-        SDL_GetRGBA(*pixel, surf->format, &pixel_r, &pixel_g, &pixel_b,     \
-                    &pixel_a);                                              \
-        tmp_r = color_r * br + pixel_r * nbr;                               \
-        tmp_g = color_g * br + pixel_g * nbr;                               \
-        tmp_b = color_b * br + pixel_b * nbr;                               \
-        tmp_a = color_a * br + pixel_a * nbr;                               \
-        *((Uint32 *)pixel) =                                                \
-            SDL_MapRGBA(surf->format, (Uint8)((tmp_r > 254) ? 255 : tmp_r), \
-                        (Uint8)((tmp_g > 254) ? 255 : tmp_g),               \
-                        (Uint8)((tmp_b > 254) ? 255 : tmp_b),               \
-                        (Uint8)((tmp_a > 254) ? 255 : tmp_a));              \
-    }                                                                       \
-    else {                                                                  \
-        pixel[0] = (Uint8)(colorptr[0] * br);                               \
-        pixel[1] = (Uint8)(colorptr[1] * br);                               \
-        pixel[2] = (Uint8)(colorptr[2] * br);                               \
-        if (hasalpha)                                                       \
-            pixel[3] = br * 255;                                            \
+static Uint32
+get_pixel_32(Uint8 *pixels, SDL_PixelFormat *format)
+{
+    switch (format->BytesPerPixel) {
+        case 4:
+            return *((Uint32*)pixels);
+        case 3:
+#if SDL_BYTEORDER == SDL_LIL_ENDIAN
+		    return *pixels | *(pixels+1) << 8 | *(pixels+2) << 16;
+#else
+            return *pixels << 16 | *(pixels+1) << 8 | *(pixels+2);
+#endif
+        case 2:
+            return *((Uint16*)pixels);
+        case 1:
+            return *pixels;
+    }
+    return 0;
+}
+
+static void
+set_pixel_32(Uint8 *pixels, SDL_PixelFormat *format, Uint32 pixel)
+{
+    switch (format->BytesPerPixel) {
```

Comment:
Cool! It looks similar to the code in **drawline** now, but why it is much more simple here? Can we use this code in *drawline* ?

Reply:
In drawline, the line is drawn for each case inside the switch statement, probably because it might be faster.